include/ublk_cmd.h

```diff
@@ -17,6 +17,8 @@
 #define	UBLK_CMD_STOP_DEV	0x07
 #define	UBLK_CMD_SET_PARAMS	0x08
 #define	UBLK_CMD_GET_PARAMS	0x09
+#define	UBLK_CMD_REPORT_ZONES_FETCH 10
+#define	UBLK_CMD_REPORT_ZONES_COMMIT_FETCH 11
```

Comment:
Here we should let SET_PARAS/GET_PARAMS to cover any zone
specific parameters. And you should define one  new structure of 'ublk_param_zoned' which includes all zoned specific definitions, meantime reserve a few fields for future extension. 

The above is the UAPI(kabi) between userspace and kernel driver,
which is important and should be done first.

Once the kabi definition is finalized, you can send one patch to linux-block for starting the zoned support, and this kind of chang shouldn't be big, just wire ublk/zoned with block layer.

Meantime working on userspace side(ublksrv side), so people
can verify the driver change. IMO, we should have one dedicated zoned target for simulating all zoned behavior, such as write on each zone, ZNS, ...

Reply:
> Here we should let SET_PARAS/GET_PARAMS to cover any zone specific parameters. And you should define one new structure of 'ublk_param_zoned' which includes all zoned specific definitions, meantime reserve a few fields for future extension.

I added zone size via the `chunk_sectors` field that is already present in the basic parameters.

> The above is the UAPI(kabi) between userspace and kernel driver, which is important and should be done first.
> 
> Once the kabi definition is finalized, you can send one patch to linux-block for starting the zoned support, and this kind of chang shouldn't be big, just wire ublk/zoned with block layer.

The kernel patch is currently undergoing internal review. I will send it next week when it is done (hopefully). The reason the patch is so big is `report_zones`. It is very few lines to wire all the zone operations in `ublk`, just add some conversion of the operation enumerations. But because `report_zones` is out-of-band in the block layer, it requires special handling. I added a uring based interface on the control device for this, leaning on how the data plane is implemented. `report_zones` is metadata, so I think it belongs in the control plane rather than the data plane.

If there is a more clean way to implement `report_zones` I can of course change the patches.

> Meantime working on userspace side(ublksrv side), so people can verify the driver change. IMO, we should have one dedicated zoned target for simulating all zoned behavior, such as write on each zone, ZNS, ...

This patch adds support to the null target and to the loopback target. The null target is not much for testing, but I have run real workloads with the loopback target backed by a ZN540 drive. I did not try SMR yet, but I will test on SMR next week.

Reply:
> > Here we should let SET_PARAS/GET_PARAMS to cover any zone specific parameters. And you should define one new structure of 'ublk_param_zoned' which includes all zoned specific definitions, meantime reserve a few fields for future extension.
> 
> I added zone size via the `chunk_sectors` field that is already present in the basic parameters.
> 

chunk_sectors can be reused for zoned device without any issue just
like kernel block layer, but other zoned parameters(max_open_zones,
max_active_zones) should be set/get by one standalone zoned parameter.

Adding them into ublk_param_basic will break kabi, and any new
parameter has to be add to the end of ublk_params. 

> > The above is the UAPI(kabi) between userspace and kernel driver, which is important and should be done first.
> > Once the kabi definition is finalized, you can send one patch to linux-block for starting the zoned support, and this kind of chang shouldn't be big, just wire ublk/zoned with block layer.
> 
> The kernel patch is currently undergoing internal review. I will send it next week when it is done (hopefully). The reason the patch is so big is `report_zones`. It is very few lines to wire all the zone operations in `ublk`, just add some conversion of the operation enumerations. But because `report_zones` is out-of-band in the block layer, it requires special handling. I added a uring based interface on the control device for this, leaning on how the data plane is implemented. `report_zones` is metadata, so I think it belongs in the control plane rather than the data plane.
> 

All control commands are supposed for handling generic control
function only, not for target specific function, that is how ublk's parameter is designed. 

I believe the zone info can be abstracted as 'struct ublk_param_zoned', then report_zones can be implemented
inside ublk driver based on 'ublk_param_zoned' as one generic
block feature.

If the approach of adding ublk_param_zoned doesn't work, please just
let us know, and I am happy to figure out better solution.

> If there is a more clean way to implement `report_zones` I can of course change the patches.
> 

Addding 'ublk_param_zoned ' is the clean approach for supporting
zoned, and adding 'ublk_param_foo' is supposed to be the normal
way to add new feature.

> > Meantime working on userspace side(ublksrv side), so people can verify the driver change. IMO, we should have one dedicated zoned target for simulating all zoned behavior, such as write on each zone, ZNS, ...
> 
> This patch adds support to the null target and to the loopback target. The null target is not much for testing, but I have run real workloads with the loopback target backed by a ZN540 drive. I did not try SMR yet, but I will test on SMR next week.

I'd suggest to not modify null/loop targets for zoned, and adding
one totally new zoned target could be better, then you can extend
it to one really zoned implementation step by step, such as, add
ZNS, ....

Thanks,
Ming

Reply:
No, please don't handle report zone by control command.
Instead, there could be at least two ways:

1) define REPORT_ZONE as one IO command, just like
other UBLK_IO_OP_ZONE_* you added, but one thing could be
that the per-io buffer may not hold all the requested zoned info
in single disk->fops->report_zones(), but the ublk driver can split
it into several requests, and each request is done by one blk-mq
pass-through request, please refer to nvme_ns_report_zones().

Then ublk server can handle them just like one normal IO.

2) store all zoned info as one PARAMETER, but this way may
waste memory a lot.

And I prefer to the 1st approach.